lib/src/install.rs

```diff
@@ -1425,6 +1486,212 @@ impl BoundImages {
+        Some(u) => u,
+        None => anyhow::bail!("Expected rootfs to have a UUID by now"),
+    };
+
+    let root_uuid_karg = format!("root=UUID={rootfs_uuid}");
+
+    let mut cmdline_refs = vec!["console=ttyS0,115200", &root_uuid_karg, "rw"];
+
+    cmdline_refs.extend(root_setup.kargs.iter().map(String::as_str));
+
+    composefs_write_boot_simple(
+        &repo,
+        entry,
+        &id,
+        root_setup.physical_root_path.as_std_path(),
+        Some("boot"),
+        Some(&format!("{}", id.to_hex())),
+        &cmdline_refs,
+    )?;
+
+    Ok(())
+}
+
+fn get_esp_device() -> Option<PathBuf> {
+    let esp_devices = [COREOS_ESP_PART_LABEL, ANACONDA_ESP_PART_LABEL]
+        .into_iter()
+        .map(|p| Path::new("/dev/disk/by-partlabel/").join(p));
+    let mut esp_device = None;
+    for path in esp_devices {
+        if path.exists() {
+            esp_device = Some(path);
+            break;
+        }
+    }
+    return esp_device;
+}
+
+/// esp_device - /dev/disk/by-partlabel/<ESP_DEVICE>
+fn get_esp_uuid(esp_device: &PathBuf) -> Result<String> {
+    // not using blkid here as the output might change from under us
+    let resolved = std::fs::canonicalize(esp_device)
+        .with_context(|| format!("Failed to resolve link {esp_device:?}"))?;
+
+    let mut uuid = String::new();
+
+    for dir_entry in std::fs::read_dir("/dev/disk/by-uuid")? {
+        let file = dir_entry?;
+
+        let uuid_resolve = std::fs::canonicalize(file.path())
+            .with_context(|| format!("Failed to resolve link {file:?}"))?;
+
+        if resolved == uuid_resolve {
+            // SAFETY: UUID has to be [A-Fa-f0-9\-]
+            uuid = file.file_name().to_string_lossy().into_owned();
+            break;
+        }
+    }
+
+    Ok(uuid)
+}
+
+#[context("Setting up UKI boot")]
+fn setup_composefs_uki_boot(
+    root_setup: &RootSetup,
+    // TODO: Make this generic
+    repo: ComposefsRepository<Sha256HashValue>,
+    id: &Sha256HashValue,
+    entry: BootEntry<Sha256HashValue>,
+) -> Result<()> {
+    // Write the UKI to <ESP>/EFI/Linux
+    let Some(esp_device) = get_esp_device() else {
+        anyhow::bail!("ESP device not found");
+    };
+
+    let mounted_esp: PathBuf = root_setup.physical_root_path.join("../esp").into();
+    create_dir_all(&mounted_esp).context("Failed to create dir {mounted_esp:?}")?;
+
+    Task::new("Mounting ESP", "mount")
+        .args([&esp_device, &mounted_esp.clone()])
+        .run()?;
+
+    composefs_write_boot_simple(
+        &repo,
+        entry,
+        &id,
+        &mounted_esp,
+        None,
+        Some(&format!("{}", id.to_hex())),
+        &[],
+    )?;
+
+    Task::new("Unmounting ESP", "umount")
+        .arg(mounted_esp)
+        .run()?;
+
+    let boot_dir = root_setup.physical_root_path.join("boot");
+    create_dir_all(&boot_dir).context("Failed to create boot dir")?;
+
+    // Add the user grug cfg
+    let grub_user_config = format!(
```

Comment:
In my ideal world grub related things live in bootupd, not in this project.

And even better if we can make this a *static* configuration...where grub basically implements the BLS spec and finds things in `/EFI/Linux` on its own right?

Reply:
This part of the GRUB config can not be generated by bootup as it is the deployment boot order in the UKI + GRUB case. See: https://github.com/containers/composefs-rs/issues/38#issuecomment-3007646048